README
XMOUNT README FILE

Table of contents
  0.0 Author and license stuff
  1.0 What is xmount? - A short description
  2.0 A deeper view of things
    2.1 Emulated DD file
    2.2 Emulated VDI file
    2.3 Emulated VMDK file
    2.4 Emulated VHD file
    2.5 Virtual write access
  3.0 Installation instructions
    3.1 Prerequisits
    3.1 Install from a package
    3.2 Install from source
  4.0 xmount command line parameters
    4.1 xmount usage examples

0.0 Author and license stuff
  xmount Copyright (c) 2008-2012 by Gillen Daniel <gillen.dan@pinguin.lu>

  This program is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with
  this program. If not, see <http://www.gnu.org/licenses/>.

  ---

  Due to AFFLIB, I have to mention that:

  This product includes software developed by Simson L. Garfinkel and Basis
  Technology Corp.

1.0 What is xmount? - A short description
  xmount allows you to convert on-the-fly between multiple input and output
  harddisk image types. xmount creates a virtual file system using FUSE
  (Filesystem in Userspace) that contains a virtual representation of the input
  image. The virtual representation can be in raw DD, VirtualBox's virtual disk
  file format, Microsoft's Virtual Hard Disk Image format or in VmWare's VMDK
  file format. Input images can be raw DD, EWF (Expert Witness Compression
  Format) or AFF (Advanced Forensic Format) files. In addition, xmount also
  supports virtual write access to the output files that is redirected to a
  cache file. This makes it possible to boot acquired harddisk images using
  QEMU, KVM, VirtualBox, VmWare or alike.

2.0 A deeper view of things
  2.1 Emulated DD file
    This is the standard data representation used by xmount. It is a file
    containing the raw data contained in the input image file.

  2.2 Emulated VDI file
    When using the "--out vdi" command line parameter, a VirtualBox disk file
    is emulated. This means that xmount builds a virtual VDI header and prepends
    it to the raw data contained in the input image. The VDI file can be used in
    VirtualBox as a virtual disk file.

  2.3 Emulated VMDK file
    When using the "--out vmdk(s)" command line paramter, a VMWare virtual disk
    file will be emulated. This mainly consists in building a VMDK descriptor
    file that will access an emulated DD file. You can use this VMDK file to
    add your input image as virtual disk to a VMWare virtual machine. "vmdk" is
    a VMWare disk attached to the IDE bus and "vmdks" is attached to the SCSI
    bus.

    When emulating VMDK files to be used in VMWare, you have to enable
    "user_allow_other" in /etc/fuse.conf or run xmount and VmWare as root.

    WARNING: VMDK support should be considered experimental and currently the
             VMDK descriptor file isn't cached.

  2.4 Emulated VHD file
    When using the "--out vhd" command line parameter, a Virtual Hard Disk Image
    file is emulated by appending to the raw data a valid VHD footer. Valid here
    means as described in Microsoft's Virtual Hard Disk Image Format specs v1.

  2.5 Virtual write access
    By using the "--cache <cache_file>" command line parameter, xmount allows
    you to change data in the emulated image files. All changes are written to a
    separate cache file. No data will ever be written to the original input
    image files! Write access is limited to change existing data. It isn't
    possible to change the emulated image's file size (no append or truncate)!
    The same cache file can be used with different output image formats without
    loosing changed data.

3.0 Installation instructions
  To install xmount, you can use a prebuild package for your distribution or
  you can build xmount from source. The two methods are described in the
  following sections.

  3.1 Prerequisits
    SYSTEM:
      Some sort of Linux with a recent kernel (2.6.x or above).
    FUSE:
      Your os must support FUSE and have the appropriate libraries installed.
      (See http://fuse.sourceforge.net/ for more informations). To be able to
      install from source, you also need the FUSE header files.
    LIBEWF:
      To enable EWF input image support. Get it from
      http://sourceforge.net/projects/libewf/
    LIBAFF:
      To enable AFF input image support. Get it from http://www.afflib.org/

  3.2 Install from a package
    Chances are I provide prebuild binary packages for Debian / Ubuntu. In this
    case, you only have to fire up your distribution's package manager and
    install xmount. See https://www.pinguin.lu for more informations and
    download links.

  3.3 Install from source
    After having checked the prerequisits, you should be able to compile and
    install xmount by simply issuing the following three commands:

    # ./configure
    # make
    # make install

4.0 xmount command line parameters
  xmount [[fopts] [mopts]] <ifile> [<ifile> [...]] <mntp>

  fopts:
    -d : Enable FUSE's and xmount's debug mode.
    -s : Run single threaded.
    -o no_allow_other : Disable automatic addition of FUSE's allow_other option.
    -o <fmopts> : Specify fuse mount options. Will also disable automatic
                  addition of FUSE's allow_other option!
  mopts:
    --cache <file> : Enable virtual write support and set cachefile to use.
    --in <itype> : Input image format. <itype> can be "dd", "ewf".
    --info : Print out some infos about used compiler and libraries.
    --out <otype> : Output image format. <otype> can be "dd", "vdi", "vhd", "vmdk(s)".
    --owcache <file> : Same as --cache <file> but overwrites existing cache.
    --rw <file> : Same as --cache <file>.
    --version : Same as --info.
    INFO: Input and output image type defaults to "dd" if not specified.
  ifile:
    Input image file. If you use EWF files, you have to specify all image
    segments! (If your shell supports it, you can use .E?? as file extension
    to specify them all)
  mntp:
    Mount point where virtual files should be located.

  4.1 xmount usage examples
    To xmount an EWF harddisk image from your acquired disk as a raw DD image
    under /mnt, use the following command:

      xmount --in ewf ./acquired_disk.E?? /mnt

    To xmount the same ewf image as vdi file, you would use a command like this:

      xmount --in ewf --out vdi ./acquired_disk.E?? /mnt

    And to enable virtual write support on a raw DD input image xmounted as VDI
    file:

      xmount --out vdi --cache ./acquired_disk.cache ./acquired_disk.dd /mnt



NEWS
This file lists the most important changes to the previous released version.
For a full list of changes please refer to the source files.

New for version 0.5.0:
  - Added support for VHD emulation.

New for version 0.4.6:
  - Added support for libewf v2

New for version 0.4.5:
  - Show the name of the mounted file in mount output

New for version 0.4.4:
  - Fixed a bug that prevents VirtualBox 3.2.8 or above from recognising the
    generated VDIs.

New for version 0.4.1:
  - Code cleanups and some minor bugfixes

New for version 0.4.0:
  - AFFLIB support for input images
  - AFFLIB and LIBEWF are now statically linked to xmount

New for version 0.3.2:
  - Automatic addition of FUSE's "-o allow_other" option if it is supported.

New for version 0.3.1:
  - Some bugfixes for VMDK emulation (Still experimental)
  - It is now possible to specify whether VMDK disks are attached to the IDE
    or SCSI bus. (Output type vmdk or vmdks)

New for version 0.3.0:
  - Added experimental support for VMDK emulation

New for version 0.2.3:
  - Virtual write support finally fixed

New for version 0.2.2:
  - Fixed yet another bug in virtual write support

New for version 0.2.1:
  - Fixed a bug in virtual write support

New for version 0.2.0:
  - Name changed from mountewf to xmount 
  - Support for multiple input image formats (DD and EWF)
  - Support for virtual write access

New for version 0.1.2:
  - Added random generator initialisation (Makes it possible to use more than 1
    mounted vdi in VBox)

New for version 0.1.1:
  - Fixed some severe bugs

New for version 0.1.0:
  - First release
  - Support for emulation of dd and vdi files


md5.h
// This file has been copied 1:1 from the source of the Debian fdupes package into Guymager.
// Only has been added this comment, no other changes.

/*
  Copyright (C) 1999 Aladdin Enterprises.  All rights reserved.

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  L. Peter Deutsch
  ghost@aladdin.com

 */
/*$Id: md5.h $ */
/*
  Independent implementation of MD5 (RFC 1321).

  This code implements the MD5 Algorithm defined in RFC 1321.
  It is derived directly from the text of the RFC and not from the
  reference implementation.

  The original and principal author of md5.h is L. Peter Deutsch
  <ghost@aladdin.com>.  Other authors are noted in the change history
  that follows (in reverse chronological order):

  1999-11-04 lpd Edited comments slightly for automatic TOC extraction.
  1999-10-18 lpd Fixed typo in header comment (ansi2knr rather than md5);
	added conditionalization for C++ compilation from Martin
	Purschke <purschke@bnl.gov>.
  1999-05-03 lpd Original version.
 */

#ifndef md5_INCLUDED
#  define md5_INCLUDED

/*
 * This code has some adaptations for the Ghostscript environment, but it
 * will compile and run correctly in any environment with 8-bit chars and
 * 32-bit ints.  Specifically, it assumes that if the following are
 * defined, they have the same meaning as in Ghostscript: P1, P2, P3,
 * ARCH_IS_BIG_ENDIAN.
 */

typedef unsigned char md5_byte_t; /* 8-bit byte */
typedef unsigned int md5_word_t; /* 32-bit word */

/* Define the state of the MD5 Algorithm. */
typedef struct md5_state_s {
    md5_word_t count[2];	/* message length in bits, lsw first */
    md5_word_t abcd[4];		/* digest buffer */
    md5_byte_t buf[64];		/* accumulate block */
} md5_state_t;

#ifdef __cplusplus
extern "C"
{
#endif

/* Initialize the algorithm. */
#ifdef P1
void md5_init(P1(md5_state_t *pms));
#else
void md5_init(md5_state_t *pms);
#endif

/* Append a string to the message. */
#ifdef P3
void md5_append(P3(md5_state_t *pms, const md5_byte_t *data, int nbytes));
#else
void md5_append(md5_state_t *pms, const md5_byte_t *data, int nbytes);
#endif

/* Finish the message and return the digest. */
#ifdef P2
void md5_finish(P2(md5_state_t *pms, md5_byte_t digest[16]));
#else
void md5_finish(md5_state_t *pms, md5_byte_t digest[16]);
#endif

#ifdef __cplusplus
}  /* end extern "C" */
#endif

#endif /* md5_INCLUDED */






xmount_cache.c
/*******************************************************************************
* xmount Copyright (c) 2008,2009 by Gillen Daniel <gillen.dan@pinguin.lu>      *
*                                                                              *
* xmount is a small tool to "fuse mount" various image formats as dd or vdi    *
* files and enable virtual write access.                                       *
*                                                                              *
* This program is free software: you can redistribute it and/or modify it      *
* under the terms of the GNU General Public License as published by the Free   *
* Software Foundation, either version 3 of the License, or (at your option)    *
* any later version.                                                           *
*                                                                              *
* This program is distributed in the hope that it will be useful, but WITHOUT  *
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        *
* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for     *
* more details.                                                                *
*                                                                              *
* You should have received a copy of the GNU General Public License along with *
* this program. If not, see <http://www.gnu.org/licenses/>.                    *
*******************************************************************************/




xmount_cache.h
/*******************************************************************************
* xmount Copyright (c) 2008,2009 by Gillen Daniel <gillen.dan@pinguin.lu>      *
*                                                                              *
* xmount is a small tool to "fuse mount" various image formats as dd or vdi    *
* files and enable virtual write access.                                       *
*                                                                              *
* This program is free software: you can redistribute it and/or modify it      *
* under the terms of the GNU General Public License as published by the Free   *
* Software Foundation, either version 3 of the License, or (at your option)    *
* any later version.                                                           *
*                                                                              *
* This program is distributed in the hope that it will be useful, but WITHOUT  *
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        *
* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for     *
* more details.                                                                *
*                                                                              *
* You should have received a copy of the GNU General Public License along with *
* this program. If not, see <http://www.gnu.org/licenses/>.                    *
*******************************************************************************/

/*
 * Define CACHE_BLOCK_SIZE to be 512KB and BLOCK_MAP_ENTRYS to 1310720.
 * This allows cache objects of max 5TB.
 */
#ifdef __LP64__
  #define CACHE_BLOCK_SIZE 524288
  #define BLOCK_MAP_ENTRYS 1310720
#else
  #define CACHE_BLOCK_SIZE 524288LL
  #define BLOCK_MAP_ENTRYS 1310720LL
#endif

/*
 * The block map of a cache object.
 */
typedef uint64_t TBlockMap;

/*
 * Different cache object types
 */
typedef enum TCacheObjectType {
  TCacheObjectType_PublicDyn,  // A public object with dynamic size
  TCacheObjectType_PublicFix,  // A public object with fixed size
  TCacheObjectType_Private     // A private object for internal use
} TCacheObjectType;

typedef struct TCacheObjectInfo {
  uint64_t ObjectSize;
}

/*
 * A cache object
 */
typedef struct TCacheObject {
  uint64_t ObjectId;       // Object identifier
  TObjectType ObjectType;  // Object type
  uint64_t offBlockMap;    // Offset to block map
} TCacheObject;

/*
 * The cache file
 */
typedef struct TCacheFile {
  uint64_t CacheBlockSize;
  TCacheObject *pCacheObjects;
} TCacheFile;

/*
 * Public cache functions
 */
int OpenCache(TCacheFile *pCacheFile,
              char *pFile);
int CloseCache(TCacheFile *pCacheFile);
int CreateCacheObject(TCacheFile *pCacheFile,
                      uint64_t ObjectId,
                      TCacheObjectType ObjectType,
                      uint64_t ObjectSize=0);
int DestroyCacheObject(TCacheFile *pCacheFile,
                       uint64_t ObjectId);
int GetCacheObjectInfo(TCacheFile *pCacheFile,
                       uint64_t ObjectId,
                       TCacheObjectInfo *pObjectInfo);
int GetCacheObjectData(TCacheFile *pCacheFile,
                       uint64_t ObjectId,
                       char *buf,
                       uint64_t offset,
                       uint64_t size);
int SetCacheObjectData(TCacheFile *pCacheFile,
                       uint64_t ObjectId,
                       char *buf,
                       uint64_t offset,
                       uint64_t size);

/*
  ----- Change history -----
  20090413: 


xmount.h
/*******************************************************************************
* xmount Copyright (c) 2008-2012 by Gillen Daniel <gillen.dan@pinguin.lu>      *
*                                                                              *
* xmount is a small tool to "fuse mount" various image formats and enable      *
* virtual write access.                                                        *
*                                                                              *
* This program is free software: you can redistribute it and/or modify it      *
* under the terms of the GNU General Public License as published by the Free   *
* Software Foundation, either version 3 of the License, or (at your option)    *
* any later version.                                                           *
*                                                                              *
* This program is distributed in the hope that it will be useful, but WITHOUT  *
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        *
* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for     *
* more details.                                                                *
*                                                                              *
* You should have received a copy of the GNU General Public License along with *
* this program. If not, see <http://www.gnu.org/licenses/>.                    *
*******************************************************************************/

#define FUSE_USE_VERSION 26
#include <fuse.h>
#include <sys/types.h>
#include <inttypes.h>
#include <stdarg.h>

#undef FALSE
#undef TRUE
#define FALSE 0
#define TRUE 1

#ifndef __APPLE__
  #define FOPEN fopen64
#else
  // Apple does use fopen for fopen64 too
  #define FOPEN fopen
#endif

/*
 * Constants
 */
#define IMAGE_INFO_HEADER "The following values have been extracted from " \
                          "the mounted image file:\n\n"

/*
 * Virtual image types
 */
typedef enum TVirtImageType {
  /** Virtual image is a DD file */
  TVirtImageType_DD,
  /** Virtual image is a DMG file */
  TVirtImageType_DMG,
  /** Virtual image is a VDI file */
  TVirtImageType_VDI,
  /** Virtual image is a VMDK file (IDE bus)*/
  TVirtImageType_VMDK,
  /** Virtual image is a VMDK file (SCSI bus)*/
  TVirtImageType_VMDKS,
  /** Virtual image is a VHD file*/
  TVirtImageType_VHD
} TVirtImageType;

/*
 * Input image types
 */
typedef enum TOrigImageType {
  /** Input image is a DD file */
  TOrigImageType_DD,
  /** Input image is an EWF file */
  TOrigImageType_EWF,
  /** Input image is an AFF file */
  TOrigImageType_AFF
} TOrigImageType;

/*
 * Various mountimg runtime options
 */
typedef struct TXMountConfData {
  /** Input image type */
  TOrigImageType OrigImageType;
  /** Virtual image type */
  TVirtImageType VirtImageType;
  /** Enable debug output */
  uint32_t Debug;
  /** Path of virtual image file */
  char *pVirtualImagePath;
  /** Path of virtual VMDK file */
  char *pVirtualVmdkPath;
  /** Path of virtual image info file */
  char *pVirtualImageInfoPath;
  /** Enable virtual write support */
  uint32_t Writable;
  /** Overwrite existing cache */
  uint32_t OverwriteCache;
  /** Cache file to save changes to */
  char *pCacheFile;
  /** Size of input image */
  uint64_t OrigImageSize;
  /** Size of virtual image */
  uint64_t VirtImageSize;
  /** Partial MD5 hash of input image */
  uint64_t InputHashLo;
  uint64_t InputHashHi;
} __attribute__ ((packed)) TXMountConfData;

/*
 * VDI Binary File Header structure
 */
#define VDI_FILE_COMMENT "<<< This is a virtual VDI image >>>"
#define VDI_HEADER_COMMENT "This VDI was emulated using xmount v" \
                           PACKAGE_VERSION
#define VDI_IMAGE_SIGNATURE 0xBEDA107F // 1:1 copy from hp
#define VDI_IMAGE_VERSION 0x00010001 // Vers 1.1
#define VDI_IMAGE_TYPE_FIXED 0x00000002 // Type 2 (fixed size)
#define VDI_IMAGE_FLAGS 0
#define VDI_IMAGE_BLOCK_SIZE (1024*1024) // 1 Megabyte
typedef struct TVdiFileHeader {
// ----- VDIPREHEADER ------
  /** Just text info about image type, for eyes only. */
  char szFileInfo[64];
  /** The image signature (VDI_IMAGE_SIGNATURE). */
  uint32_t u32Signature;
  /** The image version (VDI_IMAGE_VERSION). */
  uint32_t u32Version;
// ----- VDIHEADER1PLUS -----
  /** Size of header structure in bytes. */
  uint32_t cbHeader;
  /** The image type (VDI_IMAGE_TYPE_*). */
  uint32_t u32Type;
  /** Image flags (VDI_IMAGE_FLAGS_*). */
  uint32_t fFlags;
  /** Image comment. (UTF-8) */
  char szComment[256];
  /** Offset of Blocks array from the begining of image file.
   * Should be sector-aligned for HDD access optimization. */
  uint32_t offBlocks;
  /** Offset of image data from the begining of image file.
   * Should be sector-aligned for HDD access optimization. */
  uint32_t offData;
  /** Legacy image geometry (previous code stored PCHS there). */
  /** Cylinders. */
  uint32_t cCylinders;
  /** Heads. */
  uint32_t cHeads;
  /** Sectors per track. */
  uint32_t cSectors;
  /** Sector size. (bytes per sector) */
  uint32_t cbSector;
  /** Was BIOS HDD translation mode, now unused. */
  uint32_t u32Dummy;
  /** Size of disk (in bytes). */
  uint64_t cbDisk;
  /** Block size. (For instance VDI_IMAGE_BLOCK_SIZE.) Must be a power of 2! */
  uint32_t cbBlock;
  /** Size of additional service information of every data block.
   * Prepended before block data. May be 0.
   * Should be a power of 2 and sector-aligned for optimization reasons. */
  uint32_t cbBlockExtra;
  /** Number of blocks. */
  uint32_t cBlocks;
  /** Number of allocated blocks. */
  uint32_t cBlocksAllocated;
  /** UUID of image. */
  uint64_t uuidCreate_l;
  uint64_t uuidCreate_h;
  /** UUID of image's last modification. */
  uint64_t uuidModify_l;
  uint64_t uuidModify_h;
  /** Only for secondary images - UUID of previous image. */
  uint64_t uuidLinkage_l;
  uint64_t uuidLinkage_h;
  /** Only for secondary images - UUID of previous image's last modification. */
  uint64_t uuidParentModify_l;
  uint64_t uuidParentModify_h;
  /** Padding to get 512 byte alignment */
  uint64_t padding0;
  uint64_t padding1;
  uint64_t padding2;
  uint64_t padding3;
  uint64_t padding4;
  uint64_t padding5;
  uint64_t padding6;
} __attribute__ ((packed)) TVdiFileHeader, *pTVdiFileHeader;

//    /** The way the UUID is declared by the DCE specification. */
//    struct
//    {
//        uint32_t    u32TimeLow;
//        uint16_t    u16TimeMid;
//        uint16_t    u16TimeHiAndVersion;
//        uint8_t     u8ClockSeqHiAndReserved;
//        uint8_t     u8ClockSeqLow;
//        uint8_t     au8Node[6];
//    } Gen;

/*
 * VHD Binary File footer structure
 *
 * At the time of writing, the specs could be found here:
 *   http://www.microsoft.com/downloads/details.aspx?
 *     FamilyID=C2D03242-2FFB-48EF-A211-F0C44741109E
 *
 * Warning: All values are big-endian!
 */
// 
#ifdef __LP64__
  #define VHD_IMAGE_HVAL_COOKIE 0x78697463656E6F63 // "conectix"
#else
  #define VHD_IMAGE_HVAL_COOKIE 0x78697463656E6F63LL 
#endif
#define VHD_IMAGE_HVAL_FEATURES 0x02000000
#define VHD_IMAGE_HVAL_FILE_FORMAT_VERSION 0x00000100
#ifdef __LP64__
  #define VHD_IMAGE_HVAL_DATA_OFFSET 0xFFFFFFFFFFFFFFFF
#else
  #define VHD_IMAGE_HVAL_DATA_OFFSET 0xFFFFFFFFFFFFFFFFLL
#endif
#define VHD_IMAGE_HVAL_CREATOR_APPLICATION 0x746E6D78 // "xmnt"
#define VHD_IMAGE_HVAL_CREATOR_VERSION 0x00000500
// This one is funny! According to VHD specs, I can only choose between Windows
// and Macintosh. I'm going to choose the most common one.
#define VHD_IMAGE_HVAL_CREATOR_HOST_OS 0x6B326957 // "Win2k"
#define VHD_IMAGE_HVAL_DISK_TYPE 0x02000000
// Seconds from January 1st, 1970 to January 1st, 2000
#define VHD_IMAGE_TIME_CONVERSION_OFFSET 0x386D97E0
typedef struct TVhdFileHeader {
  uint64_t cookie;
  uint32_t features;
  uint32_t file_format_version;
  uint64_t data_offset;
  uint32_t creation_time;
  uint32_t creator_app;
  uint32_t creator_ver;
  uint32_t creator_os;
  uint64_t size_original;
  uint64_t size_current;
  uint16_t disk_geometry_c;
  uint8_t disk_geometry_h;
  uint8_t disk_geometry_s;
  uint32_t disk_type;
  uint32_t checksum;
  uint64_t uuid_l;
  uint64_t uuid_h;
  uint8_t saved_state;
  char Reserved[427];
} __attribute__ ((packed)) TVhdFileHeader, *pTVhdFileHeader;

/*
 * Cache file block index array element
 */
#ifdef __LP64__
  #define CACHE_BLOCK_FREE 0xFFFFFFFFFFFFFFFF
#else
  #define CACHE_BLOCK_FREE 0xFFFFFFFFFFFFFFFFLL 
#endif
typedef struct TCacheFileBlockIndex {
  /** Set to 1 if block is assigned (This block has data in cache file) */
  uint32_t Assigned;
  /** Offset to data in cache file */
  uint64_t off_data;
} __attribute__ ((packed)) TCacheFileBlockIndex, *pTCacheFileBlockIndex;

/*
 * Cache file header structures
 */
#define CACHE_BLOCK_SIZE (1024*1024) // 1 megabyte
#ifdef __LP64__
  #define CACHE_FILE_SIGNATURE 0xFFFF746E756F6D78 // "xmount\xFF\xFF"
#else
  #define CACHE_FILE_SIGNATURE 0xFFFF746E756F6D78LL 
#endif
#define CUR_CACHE_FILE_VERSION 0x00000002 // Current cache file version
#define HASH_AMOUNT (1024*1024)*10 // Amount of data used to construct a
                                   // "unique" hash for every input image
                                   // (10MByte)
// Current header
typedef struct TCacheFileHeader {
  /** Simple signature to identify cache files */
  uint64_t FileSignature;
  /** Cache file version */
  uint32_t CacheFileVersion;
  /** Cache block size */
  uint64_t BlockSize;
  /** Total amount of cache blocks */
  uint64_t BlockCount;
  /** Offset to the first block index array element */
  uint64_t pBlockIndex;
  /** Set to 1 if VDI file header is cached */
  uint32_t VdiFileHeaderCached;
  /** Offset to cached VDI file header */
  uint64_t pVdiFileHeader;
  /** Set to 1 if VMDK file is cached */
  uint32_t VmdkFileCached;
  /** Size of VMDK file */
  uint64_t VmdkFileSize;
  /** Offset to cached VMDK file */
  uint64_t pVmdkFile;
  
  /** Set to 1 if VHD header is cached */
  uint32_t VhdFileHeaderCached;
  /** Offset to cached VHD header */
  uint64_t pVhdFileHeader;
  
  /** Padding until offset 512 to ease further additions */
  char HeaderPadding[432];
} __attribute__ ((packed)) TCacheFileHeader, *pTCacheFileHeader;

// Old v1 header
typedef struct TCacheFileHeader_v1 {
  /** Simple signature to identify cache files */
  uint64_t FileSignature;
  /** Cache file version */
  uint32_t CacheFileVersion;
  /** Total amount of cache blocks */
  uint64_t BlockCount;
  /** Offset to the first block index array element */
  uint64_t pBlockIndex;
  /** Set to 1 if VDI file header is cached */
  uint32_t VdiFileHeaderCached;
  /** Offset to cached VDI file header */
  uint64_t pVdiFileHeader;
  /** Set to 1 if VMDK file is cached */
} TCacheFileHeader_v1, *pTCacheFileHeader_v1;

/*
 * Macros to ease debugging and error reporting
 */
#define LOG_ERROR(...) \
  LogMessage("ERROR",(char*)__FUNCTION__,__LINE__,__VA_ARGS__);
#define LOG_WARNING(...) \
  LogMessage("WARNING",(char*)__FUNCTION__,__LINE__,__VA_ARGS__);
#define LOG_DEBUG(...) { \
  if(XMountConfData.Debug) \
    LogMessage("DEBUG",(char*)__FUNCTION__,__LINE__,__VA_ARGS__); \
}

/*
 * Macros to alloc or realloc memory and check whether it worked
 */
#define XMOUNT_MALLOC(var,var_type,size) { \
  (var)=(var_type)malloc(size); \
  if((var)==NULL) { \
    LOG_ERROR("Couldn't allocate memmory!\n"); \
    exit(1); \
  } \
}
#define XMOUNT_REALLOC(var,var_type,size) { \
  (var)=(var_type)realloc((var),size); \
  if((var)==NULL) { \
    LOG_ERROR("Couldn't allocate memmory!\n"); \
    exit(1); \
  } \
}

/*
 * Macros for some often used string functions
 */
#define XMOUNT_STRSET(var1,var2) { \
  XMOUNT_MALLOC(var1,char*,strlen(var2)+1) \
  strcpy(var1,var2); \
}
#define XMOUNT_STRNSET(var1,var2,size) { \
  XMOUNT_MALLOC(var1,char*,(size)+1) \
  strncpy(var1,var2,size); \
  (var1)[size]='\0'; \
}
#define XMOUNT_STRAPP(var1,var2) { \
  XMOUNT_REALLOC(var1,char*,strlen(var1)+strlen(var2)+1) \
  strcpy((var1)+strlen(var1),var2); \
}
#define XMOUNT_STRNAPP(var1,var2,size) { \
  XMOUNT_REALLOC(var1,char*,strlen(var1)+(size)+1) \
  (var1)[strlen(var1)+(size)]='\0'; \
  strncpy((var1)+strlen(var1),var2,size); \
}

/*
 * Macros for endian conversions
 */
// First we need to have the bswap functions
#if HAVE_BYTESWAP_H
  #include <byteswap.h>
#elif defined(HAVE_ENDIAN_H)
  #include <endian.h>
#elif defined(__APPLE__)
  #include <libkern/OSByteOrder.h>
  #define bswap_16 OSSwapInt16
  #define bswap_32 OSSwapInt32
  #define bswap_64 OSSwapInt64
#else
  #define	bswap_16(value) {                    \
    ((((value) & 0xff) << 8) | ((value) >> 8)) \
  }
  #define	bswap_32(value)	{                                     \
 	  (((uint32_t)bswap_16((uint16_t)((value) & 0xffff)) << 16) | \
 	  (uint32_t)bswap_16((uint16_t)((value) >> 16)))              \
 	}
  #define	bswap_64(value)	{                                         \
 	  (((uint64_t)bswap_32((uint32_t)((value) & 0xffffffff)) << 32) | \
 	  (uint64_t)bswap_32((uint32_t)((value) >> 32)))                  \
 	}
#endif
// Next we need to know what endianess is used
#if defined(__LITTLE_ENDIAN__)
  #define XMOUNT_BYTEORDER_LE
#elif defined(__BIG_ENDIAN__)
  #define XMOUNT_BYTEORDER_BE
#elif defined(__BYTE_ORDER__)
  #if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
    #define XMOUNT_BYTEORDER_LE
  #else
    #define XMOUNT_BYTEORDER_BE
  #endif
#endif
// And finally we can define the macros
#ifdef XMOUNT_BYTEORDER_LE
  #ifndef be16toh
    #define be16toh(x) bswap_16(x)
  #endif
  #ifndef htobe16
    #define htobe16(x) bswap_16(x)
  #endif
  #ifndef be32toh
    #define be32toh(x) bswap_32(x)
  #endif
  #ifndef htobe32
    #define htobe32(x) bswap_32(x)
  #endif
  #ifndef be64toh
    #define be64toh(x) bswap_64(x)
  #endif
  #ifndef htobe64
    #define htobe64(x) bswap_64(x)
  #endif
  #ifndef le16toh
    #define le16toh(x) (x)
  #endif
  #ifndef htole16
    #define htole16(x) (x)
  #endif
  #ifndef le32toh
    #define le32toh(x) (x)
  #endif
  #ifndef htole32
    #define htole32(x) (x)
  #endif
  #ifndef le64toh
    #define le64toh(x) (x)
  #endif
  #ifndef htole64
    #define htole64(x) (x)
  #endif
#else
  #ifndef be16toh
    #define be16toh(x) (x)
  #endif
  #ifndef htobe16
    #define htobe16(x) (x)
  #endif
  #ifndef be32toh
    #define be32toh(x) (x)
  #endif
  #ifndef htobe32
    #define htobe32(x) (x)
  #endif
  #ifndef be64toh
    #define be64toh(x) (x)
  #endif
  #ifndef htobe64
    #define htobe64(x) (x)
  #endif
  #ifndef le16toh
    #define le16toh(x) bswap_16(x)
  #endif
  #ifndef htole16
    #define htole16(x) bswap_16(x)
  #endif
  #ifndef le32toh
    #define le32toh(x) bswap_32(x)
  #endif
  #ifndef htole32
    #define htole32(x) bswap_32(x)
  #endif
  #ifndef le64toh
    #define le64toh(x) bswap_64(x)
  #endif
  #ifndef htole64
    #define htole64(x) bswap_64(x)
  #endif
#endif

/*
  ----- Change history -----
  20090226: * Added change history information to this file.
            * Added TVirtImageType enum to identify virtual image type.
            * Added TOrigImageType enum to identify input image type.
            * Added TMountimgConfData struct to hold various mountimg runtime
              options.
            * Renamed VDIFILEHEADER to TVdiFileHeader.
  20090228: * Added LOG_ERROR and LOG_DEBUG macros
            * Added defines for various static VDI header values
            * Added defines for TRUE and FALSE
  20090307: * Added defines for various static cache file header values
            * Added VdiFileHeaderCached and pVdiFileHeader values to be able to
              cache the VDI file header separatly.
  20090519: * Added new cache file header structure and moved old one to
              TCacheFileHeader_v1.
            * New cache file structure includes VmdkFileCached and pVmdkFile to
              cache virtual VMDK file and makes room for further additions so
              current cache file version 2 cache files can be easily converted
              to newer ones.
  20090814: * Added XMOUNT_MALLOC and XMOUNT_REALLOC macros.
  20090816: * Added XMOUNT_STRSET, XMOUNT_STRNSET, XMOUNT_STRAPP and
              XMOUNT_STRNAPP macros.
  20100324: * Added "__attribute__ ((packed))" to all header structs to prevent
              different sizes on i386 and amd64.
  20111109: * Added TVirtImageType_DMG type.
  20120130: * Added LOG_WARNING macro.
  20120507: * Added TVhdFileHeader structure.
  20120511: * Added endianess conversation macros
*/
